### ISSUE TITLE
Session Proposal: JavaScript modules in TC39 and Node: past, present and future

### ISSUE THREAD

**Issue by littledan** [openjs-foundation/summit#397]
### Proposal

**Topic of the session**

We all like ESM, right? But there's so much going on, between TC39 and Node.js, that it's hard to follow all of the semantics and future possibilities. Let's review where we are, to understand where we could go next.

**Type of the session**

- [x] Collaborate
- [ ] Workshop
- [ ] Talk

A group discussion

**Estimated duration of the session**

1 hour or less

**Date and Time of the session**

**Level**

- [x] Beginner
- [ ] Intermediate
- [ ] Advanced

**Pre-requisite knowledge**

Participants should be familiar with the basic syntax of ESM and CJS, and the very basics of their semantics (e.g., named exports in ESM vs an exports object in CJS). Beyond that, the session will try to be self-contained, since the purpose is to share knowledge, not test it.

**Describe the session**

- Review where we are, in terms of TC39 proposals since ES6, the Node module loader, and ecosystem adoption
- Discuss ongoing efforts in TC39 and Node for future module features
- Chat about current problems in adopting modules, whether these are addressed by those efforts or not

**Session facilitator(s), Github handle(s) and timezone(s)**

@littledan + co-facilitator wanted!

**Meeting notes and Virtual Meeting Link**

**Follow-up / Set-up sessions (if any)**

**Additional context (optional)**

----

**Comment by darcyclarke** [comment]
Just wanted to note I'd like to attend this virtually if possible. My timezone is EST.

**Comment by GeoffreyBooth** [comment]
Me too, I’m Pacific.

**Comment by joyeecheung** [comment]
For the remote participants: we've scheduled Zoom Webinars for the sessions, please register using the links provided in https://github.com/openjs-foundation/summit/issues/387 and you'll get a link in an email to join the sessions. More info in the issue mentioned.

**Comment by GeoffreyBooth** [comment]
@littledan is the intent for this session to focus on syntax stuff like TC39 proposals, or is the scope broader to include things like the Node module customization hooks?

If the latter, these two issues are top of mind related to module customizations:
- https://github.com/nodejs/node/issues/52219 could provide a migration path from monkey-patching to an official API
- https://github.com/nodejs/node/pull/50752 is the last bug that needs fixing in order to mark the current hooks API as stable

**Comment by littledan** [comment]
The idea is definitely to be broad scope and include current Node issues. If you can present on those, it would be great!

**Comment by joyeecheung** [comment]
Closing as the session has ended.

Links to the recordings:

- Day 1 morning: https://zoom.us/rec/share/UmLZkxbAqBGgEUh8GptGvS_5qXXxKqEpT-rcTMX4ncTTS715-OTIJFv4nLZ40Tj5.Q4gNxJPbpTLH0W-X?startTime=1712132193000
- Day 1 afternoon: https://zoom.us/rec/share/UmLZkxbAqBGgEUh8GptGvS_5qXXxKqEpT-rcTMX4ncTTS715-OTIJFv4nLZ40Tj5.Q4gNxJPbpTLH0W-X?startTime=1712148446000
- Day 2 morning: https://zoom.us/rec/share/UcbGphrblwjuE2jL9dlsv-4hMv6uC0M_dS96N1xByyJcXfELAvrakmjHNguiS-Nc.5wjMJg55LA891PxO?startTime=1712218676000
- Day 2 afternoon: https://zoom.us/rec/share/UcbGphrblwjuE2jL9dlsv-4hMv6uC0M_dS96N1xByyJcXfELAvrakmjHNguiS-Nc.5wjMJg55LA891PxO?startTime=1712234786000

Links to the notes (very inaccurate, we should try harder at getting dedicated note takers the next time):

- Day 1: https://hackmd.io/erevKHb6RVihtWpdPgFWCw
- Day 2: https://hackmd.io/poOBiqFiTBeCIlfOv5HwUQ

Drafted trip report, would appreciate some reviews. Expect to publish on the official blog next week: https://hackmd.io/5vvP8o5bTmqcbNh-3oHpag
